docs/best-practices/host-custom-binary-cache-with-s3.md

```diff
@@ -0,0 +1,267 @@
+# Host Custom Binary Cache with S3 {#host-custom-binary-cache-with-s3}
+
+## TL;DR
+
+A guide on how to set up your own S3 nix binary cache using MinIO S3 server.
+
+## How Software Stored in Nix? {#how-software-stored-in-nix}
+
+Multiple versions of the same software package can be installed on a system, making it
+possible to satisfy various dependency chains at a time. This enables the installation of
+multiple packages that depend on the same third package, but on different versions of it.
+To achieve this, all packages are installed in the global nix store under `/nix/store/`
+and are then symlinked to the respective locations. To verify a package's uniqueness, its
+whole directory is hashed, and the hash put into the name of the package's main folder.
```

Comment:
Not only `its whole directory`, but the whole derivation's data-structure is used to derive its hash.
This is so called the “input-addressed” model, since the output paths are computed from all the input of the build.

Details: <https://github.com/NixOS/rfcs/blob/master/rfcs/0062-content-addressed-paths.md#detailed-design>

Reply:
BTW, adding an example of the store paths here will make it more clear, e.g.

```bash
# List some store paths in /nix/store
$ ls /nix/store/ | head

001gp43bjqzx60cg345n2slzg7131za8-nix-nss-open-files.patch
001im7qm8achbyh0ywil6hif6rqf284z-bootstrap-stage0-binutils-wrapper-boot.drv
0029mr9vdfdkd3wzh6785w578j179m0l-pkg-config-wrapper-0.29.2.drv
0039201phg77q8n7djd530f46qq259yh-Librpcsvc-26.tar.gz.drv
0039fczca4mmdvs16v278q09njzibz9n-python3.11-gevent-22.10.2.drv
003bv9i3pipypw9zwcfg1s7wqa5g0ggw-fmtutil.sed
003ri07b1gfl0v2svxlapk5yjdrr0zhn-System.Reflection.Emit.4.0.1.nupkg.drv
003wkhdk1ph54r6gx6lxax53bpf0bf4s-aws-c-sdkutils-0.1.12.drv
004z9fdf9jqg29pgwh17a0qkqv7xwb3j-utf8-string-1.0.2.tar.gz.drv
0053frndrzncxwzyxffr9lnwx4kngvl5-texlive-fullwidth-0.1.drv
```

Reply:
> Not only `its whole directory`, but the whole derivation's data-structure is used to derive its hash. This is so called the “input-addressed” model, since the output paths are computed from all the input of the build.
> 
> Details: https://github.com/NixOS/rfcs/blob/master/rfcs/0062-content-addressed-paths.md#detailed-design

I don't quite get what you mean here. Could you please propose your changes directly?

Reply:
> BTW, adding an example of the store paths here will make it more clear, e.g.
> 
> ```shell
> # List some store paths in /nix/store
> $ ls /nix/store/ | head
> 
> 001gp43bjqzx60cg345n2slzg7131za8-nix-nss-open-files.patch
> 001im7qm8achbyh0ywil6hif6rqf284z-bootstrap-stage0-binutils-wrapper-boot.drv
> 0029mr9vdfdkd3wzh6785w578j179m0l-pkg-config-wrapper-0.29.2.drv
> 0039201phg77q8n7djd530f46qq259yh-Librpcsvc-26.tar.gz.drv
> 0039fczca4mmdvs16v278q09njzibz9n-python3.11-gevent-22.10.2.drv
> 003bv9i3pipypw9zwcfg1s7wqa5g0ggw-fmtutil.sed
> 003ri07b1gfl0v2svxlapk5yjdrr0zhn-System.Reflection.Emit.4.0.1.nupkg.drv
> 003wkhdk1ph54r6gx6lxax53bpf0bf4s-aws-c-sdkutils-0.1.12.drv
> 004z9fdf9jqg29pgwh17a0qkqv7xwb3j-utf8-string-1.0.2.tar.gz.drv
> 0053frndrzncxwzyxffr9lnwx4kngvl5-texlive-fullwidth-0.1.drv
> ```

Same as above.